### PR TITLE
fix(build): Patch and test dynamic linked binaries

### PR DESCRIPTION
**Pull request by dcarley**
## Proposed Changes

Patch ELF binaries from impure builds to use an interpreter from the Nix store. This is the best effort that we can make when an impure build uses toolchains and libraries from the host OS rather than the Flox environment.

Any dynamically linked libraries are resolved by unqualified name at runtime. So we would use `libc` from the Flox environment, if it provides it, in preference over the system. This order can be influenced by setting `rpath` and/or `runpath` on the binary BUT to do so for an impure build we'd have to retrospectively identify the lib/pkg names (which is hard) and add them to the closure (which is too late).

This is tested with a Go environment that doesn't contain `gcc` or `glibc` but explicitly forces CGO compilation and uses `stdio`. This is made complicated by the `devShell` as explained below and in the comments.

The `autoPatchelfHook` needs to be called explicitly because `runCommand` doesn't call any phases/hooks. `autoPatchelfPostFixup` generates a bit of noise in the build output but we can deal with that in either ticket #2044 or #2127.

---

When testing the same project outside of the `devShell` from `main` on a Ubuntu container, no `rpath` or `runpath` are set:

    root@81430f57e252:/mnt# FLOX_FEATURES_BUILD=true nix run github:flox/flox -- build -d tmp
    ✨ Build completed successfully

    root@81430f57e252:/mnt# chrpath -l tmp/result-hello/bin/.hello-wrapped
    tmp/result-hello/bin/.hello-wrapped: no rpath or runpath tag found.

Resolves `libc` from the system:

    root@81430f57e252:/mnt# ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffffbe4a4000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffbe2a0000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffffbe467000)

Same when inside the environment which doesn't contain `libc`:

    root@81430f57e252:/mnt# nix run github:flox/flox -- activate -d tmp -- ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffffa7a29000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffa7640000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffffa79ec000)

Resolves `libc` from the environment if it contains it:

    root@81430f57e252:/mnt# nix run github:flox/flox -- install glibc -d tmp
    ⚠️  Installing 'glibc' for the following systems: ["aarch64-linux", "x86_64-linux"]
    ✅ 'glibc' installed to environment 'tmp'
    root@81430f57e252:/mnt# nix run github:flox/flox -- activate -d tmp -- ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffff992fa000)
        libpthread.so.0 => /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/libpthread.so.0 (0x0000ffff99090000)
        libc.so.6 => /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/libc.so.6 (0x0000ffff98ec0000)
        /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/ld-linux-aarch64.so.1 => /nix/store/mvsq4z2g9l05vfwdkr9nihbxyxm18nak-glibc-2.39-52/lib/ld-linux-aarch64.so.1 (0x0000ffff992bd000)

---

The modified behaviour for ELF is as follows, no `rpath` or `runpath` are set, as before:

    root@81430f57e252:/mnt# FLOX_FEATURES_BUILD=true nix run .#flox -- build -d tmp
    ✨ Build completed successfully

    root@81430f57e252:/mnt# chrpath -l tmp/result-hello/bin/.hello-wrapped
    tmp/result-hello/bin/.hello-wrapped: no rpath or runpath tag found.

The interpreter is now patched but `libc` will continue to be resolved from either the Flox environment or the host OS, in that order:

    root@81430f57e252:/mnt# ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffffa32b6000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffa30b0000)
        /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/ld-linux-aarch64.so.1 => /lib/ld-linux-aarch64.so.1 (0x0000ffffa3279000)

    root@81430f57e252:/mnt# nix run github:flox/flox -- activate -d tmp -- ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffff9b295000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffff9aeb0000)
        /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/ld-linux-aarch64.so.1 => /lib/ld-linux-aarch64.so.1 (0x0000ffff9b258000)

    root@81430f57e252:/mnt# nix run github:flox/flox -- install glibc -d tmp
    ⚠️  Installing 'glibc' for the following systems: ["aarch64-linux", "x86_64-linux"]
    ✅ 'glibc' installed to environment 'tmp'
    root@81430f57e252:/mnt# nix run github:flox/flox -- activate -d tmp -- ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffffbc67d000)
        libc.so.6 => /nix/store/mvsq4z2g9l05vfwdkr9nihbxyxm18nak-glibc-2.39-52/lib/libc.so.6 (0x0000ffffbc270000)
        /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/ld-linux-aarch64.so.1 => /nix/store/mvsq4z2g9l05vfwdkr9nihbxyxm18nak-glibc-2.39-52/lib/ld-linux-aarch64.so.1 (0x0000ffffbc640000)

---

When run with `nix develop` it inherits `rpath`/`runpath` from the `devShell` which influences `ldd` from outside:

    root@81430f57e252:/mnt# FLOX_FEATURES_BUILD=true nix develop -c -- nix run github:flox/flox -- build -d tmp
    ✨ Build completed successfully

    root@81430f57e252:/mnt# chrpath -l tmp/result-hello/bin/.hello-wrapped
    tmp/result-hello/bin/.hello-wrapped: RUNPATH=/mnt/outputs/out/lib:/nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib:/nix/store/djqcjdx73hlg4ap96s478snydhpggq3j-gcc-13.3.0-lib/lib

    root@81430f57e252:/mnt# ldd tmp/result-hello/bin/.hello-wrapped
        linux-vdso.so.1 (0x0000ffffa7018000)
        libpthread.so.0 => /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/libpthread.so.0 (0x0000ffffa6fb0000)
        libc.so.6 => /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/libc.so.6 (0x0000ffffa6de0000)
        /nix/store/1c5f3l0rk178ipam8pw24dd20lv1v5a7-glibc-2.39-52/lib/ld-linux-aarch64.so.1 => /lib/ld-linux-aarch64.so.1 (0x0000ffffa6fdb000)

## Release Notes

N/A until builds general availability.

### CODE REVIEW

File: cli/flox-rust-sdk/src/providers/build.rs [hunk 1]
```diff
@@ -1010,6 +1010,127 @@ mod tests {
...
+            // Linux
+            goblin::Object::Elf(elf) => {
+                // Should always come from the Flox environment.
+                // TODO: Gives false-positives when run from devShell.
+                let interpreter = elf.interpreter.unwrap();
+                assert!(
+                    interpreter.starts_with("/nix/store/"),
+                    "should patch interpreter to nix store, got {:?}",
+                    interpreter
+                );
```
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/build.rs:1088]: Open to ideas about how to make this more reliable.

File: package-builder/build-manifest.nix
Comment (dcarley) [inline package-builder/build-manifest.nix:1]: > and add them to the closure (which is too late)

I may have misunderstood this. If we do something like `patchelf --shrink-rpath --allowed-rpath-prefixes ${pkgs.glibc}/lib` will it automatically get added to the closure, even if it wasn't a `buildInput`?

> retrospectively identify the lib/pkg names (which is hard)

This still stands though. As far as I can tell we'd need our own heuristics to find `libc` and then how far do we go in identifying other lib/pkg mappings?

Comment (limeytexan) [reply package-builder/build-manifest.nix:1]: > ... If we do something like `patchelf --shrink-rpath --allowed-rpath-prefixes ${pkgs.glibc}/lib` will it automatically get added to the closure, even if it wasn't a buildInput?

I've lost the context for these comments but the question is easily answered by building the package and running `nix-store -qR` on the result, and yes It should automatically get added to the closure because of the package scanning that takes place at the conclusion of the build.

Also, I suspect that `pkgs.glibc` won't always be the right thing to be adding in all instances. For example, consider the case where you are building with a custom version of `go` that has been linked against a modified `glibc`. But in general yes I think this is headed in the right direction.

> This still stands though ...

Can you point me to the conversation context .. can't find it anywhere in this PR???

Comment (limeytexan) [reply package-builder/build-manifest.nix:1]: Oh .. you were commenting on this entire paragraph:

> Any dynamically linked libraries are resolved by unqualified name at runtime. So we would use libc from the Flox environment, if it provides it, in preference over the system. This order can be influenced by setting rpath and/or runpath on the binary BUT to do so for an impure build we'd have to retrospectively identify the lib/pkg names (which is hard) and add them to the closure (which is too late).

This is straying into the realm of the role of `LD_AUDIT` in providing a mapping to dynamic libraries in general. This issue is about using patchelf to set the interpreter specifically, and that of course is inexorably linked to the resolution of `glibc`, but that's where it ends. After that, `LD_AUDIT` will use the `FLOX_ENV_LIB_DIRS` variable to find the best match when resolving dynamic libraries.

Bottom line I think this PR is good to go.  👍

**Review by limeytexan — APPROVED** [review]
LGTM

File: cli/flox-rust-sdk/Cargo.toml [hunk 2]
```diff
@@ -46,6 +46,7 @@ serial_test = { workspace = true, features = ["file_locks"] }
 
 [dev-dependencies]
 anyhow.workspace = true
+goblin = "0.8.2"
```
Comment (ysndr) [inline cli/flox-rust-sdk/Cargo.toml:49]: for consistency probably worth to declare this in the workspace as well

File: cli/flox-rust-sdk/src/providers/build.rs [hunk 3]
```diff
@@ -1010,6 +1010,127 @@ mod tests {
...
+    #[test]
+    fn build_impure_patches_binaries() {
+        let package_name = String::from("foo");
+        let file_name = String::from("links-against-libc");
+
+        // Force CGO without providing `gcc` from the Flox environment.
+        //
+        // In reality `gcc` and `libc` will come from the `devShell` rather than
+        // a host OS which complicates these tests and means that we can never
+        // trust `rpath` or `runpath`.
```
Comment (ysndr) [inline cli/flox-rust-sdk/src/providers/build.rs:1022]: Not sure if I understand the trust issue here.
Is this influenced by environment vars and can we perhaps make this more reliable by using `temp_env` then?

Comment (dcarley) [reply cli/flox-rust-sdk/src/providers/build.rs:1022]: "trust" might have been too stronger word and it's more of an issue for the interpreter than rpath/runpath: https://github.com/flox/flox/pull/2234#discussion_r1797111308

Any time that `gcc` and `libc` come Nix (via devShell or profile, rather than some other Linux system package) then the interpreter and rpath/runpath will point at store paths. That makes it impossible to observe the effect of the `autoPatchelfHook` by checking whether they point at store paths.

I tried unsetting `NIX_ENFORCE_NO_NATIVE`, `NIX_LDFLAGS`, and `NIX_CC` to no effect. I suspect that we'd need to revert `PATH` to prior to the `devShell` but then that would cause the tests to fail on our builders which don't have a system-wide `gcc` and `glibc`.

File: package-builder/build-manifest.nix
Comment (dcarley) [reply package-builder/build-manifest.nix:1]: Thanks! I recall there were some past discussions about using `LD_AUDIT` to warn about the use of non-portable system libs. Do you know if we have a ticket for it?

File: cli/flox-rust-sdk/src/providers/build.rs (hunk 3 above)
Comment (ysndr) [reply cli/flox-rust-sdk/src/providers/build.rs:1022]: so,.. the issue is that go uses the c compiler it finds on the PATH which may be a system provided one, but while nix wraps the compiler with additional nix flags, when running the test within a nix shell, we don't see this fail?

I wonder if putting gcc-unwrapped on top of the PATH would help, but i figure that just fails to compile without also pointing to the right stdlibs

Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/build.rs:1022]: > I suspect that we'd need to revert PATH to prior to the devShell but then that would cause the tests to fail on our builders which don't have a system-wide gcc and glibc

One way to ensure a consistent experience would be to reset the PATH to contain _only_ the build environment as we invoke the build script. That would of course impose the requirement that gcc be installed to that build environment, but is that not what we're looking for?

(Relatedly, I've been chatting with @stahnma about adding a `flox activate --pure` flag which would set the PATH to be only the flox-provided bin and sbin directories. If we had this flag then we could add it to the build script invocation on line 202 of flox-build.mk.)

Comment (ysndr) [reply cli/flox-rust-sdk/src/providers/build.rs:1022]: yeah doing impure builds and expecting a pure result dont mix too well...

> One way to ensure a consistent experience would be to reset the PATH to contain only the build environment as we invoke the build script. That would of course impose the requirement that gcc be installed to that build environment, but is that not what we're looking for?

It's certainly what I would be looking for but impure builds seem to explicitly allow for that.
A `pure` flag would be nice but i suspect that would rub badly with the semantics of `sandbox = "pure"`.

File: package-builder/build-manifest.nix [hunk 4]
```diff
@@ -64,6 +65,9 @@ pkgs.runCommandNoCC name
             ${pkgs.lib.optionalString pkgs.stdenv.isDarwin ''
               signDarwinBinariesInAllOutputs
             ''}
+            ${pkgs.lib.optionalString pkgs.stdenv.isLinux ''
+              autoPatchelfPostFixup
```
Comment (mkenigs) [inline package-builder/build-manifest.nix:69]: I'm still catching up on what this fixes...I see in the commit message that the interpreter is patched, but since it still resolves the same way in all the situations you showed, why do we care?

File: cli/flox-rust-sdk/src/providers/build.rs (hunk 3 above)
Comment (mkenigs) [reply cli/flox-rust-sdk/src/providers/build.rs:1022]: Should we skip the test if `gcc` is coming from `/nix/store`?

File: cli/flox-rust-sdk/src/providers/build.rs (hunk 1 above)
Comment (mkenigs) [reply cli/flox-rust-sdk/src/providers/build.rs:1089]: Could we differentiate between different libcs by using the full store path?

Comment (dcarley) [reply cli/flox-rust-sdk/src/providers/build.rs:1089]: I'm not sure how this would help. How would libc come from a different store path than the devShell?

File: package-builder/build-manifest.nix (hunk 4 above)
Comment (dcarley) [reply package-builder/build-manifest.nix:69]: It provides the slim guarantee that the interpreter/linker (ld-linux-aarch64) comes from a consistent place but I'm not sure that it makes any practical difference. We can't do what the ticket originally asked for, which is to find store path equivalents of any system libraries that an impure build might have linked against.

File: cli/flox-rust-sdk/src/providers/build.rs (hunk 1 above)
Comment (mkenigs) [reply cli/flox-rust-sdk/src/providers/build.rs:1089]: Still not completely sure I have my head wrapped around this. But if I tweak the test to install gcc@wrapper-12.2.0, revert the `autoPatchelfPostFixup`, and make this `assert!` and `assert_eq!` just to show what store path it is, I see `/nix/store/yaz7pyf0ah88g2v505l38n0f3wg2vzdj-glibc-2.37-8/lib/ld-linux-x86-64.so.2`

If I then undo reverting the patch, I see `/nix/store/m71p7f0nymb19yn1dascklyya2i96jfw-glibc-2.39-52/lib/ld-linux-x86-64.so.2`

So I think in the first case it's coming from gcc@wrapper-12.2.0, and in the second we're unconditionally patching to whatever is in nixpkgs rev `dc14ed...` pinned in build-manifest.nix

Looks like the devShell would probably bring the same as `dc14ed...` until https://github.com/flox/flox/pull/2231 (sidenote, should I have bumped the pin in that PR?)

We'll need to stop pinning nixpkgs in build-manifest.nix and instead use the same nixpkgs as the toplevel group (https://github.com/flox/flox/issues/2084), so maybe we could change the test to:
- Install gcc@old-version in package group other than toplevel (which prior to this patch would cause interpreter to be set to `/nix/store/yaz7pyf0ah88g2v505l38n0f3wg2vzdj-glibc-2.37-8/lib/ld-linux-x86-64.so.2`)
- Assert that interpreter is `/nix/store/m71p7f0nymb19yn1dascklyya2i96jfw-glibc-2.39-52/lib/ld-linux-x86-64.so.2` (which is correct as of this patch)
- When we do https://github.com/flox/flox/issues/2084, we can assert that the interpreter is set to whatever the toplevel group provides

Comment (dcarley) [reply cli/flox-rust-sdk/src/providers/build.rs:1089]: That sounds like it's doing the opposite of what we wanted to achieve by no longer using gcc/glibc from the environment when present. In fact, if I include `gcc@wrapper-12.2.0` and `autoPatchelfPostFixup`, then I end up with a binary that just hangs. So I think we:

1. shouldn't add `autoPatchelfPostFixup`
2. should the portability risks of impure builds in the documentation
3. can explore an `LD_AUDIT` warn mode separately
4. could still add a test to verify that a binary built with Go and GCC works as a regression test against adding `autoPatchelfPostFixup` but don't bother inspecting any attributes of the unwrapped binary

**Comment by dcarley** [comment]
Closing without merging based on this comment: https://github.com/flox/flox/pull/2234#discussion_r1815168910

Will open a separate PR for the standalone test described in bullet point 4.
